### PR TITLE
Generate plain text corpus via export command

### PR DESCRIPTION
**Pull request by quadrismegistus**
For #556

### CODE REVIEW

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+
+@contextmanager
+def logging_disabled(highest_level=logging.CRITICAL):
+    """Quick way to suppress solr logs as we iterate. Taken from https://gist.github.com/simon-weber/7853144"""
+    previous_level = logging.root.manager.disable
+    logging.disable(highest_level)
+    try: yield
+    finally: logging.disable(previous_level)
+
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:174]: This should be handled by updating django logging config rather than overriding here (we might need to turn logging on to debug something). What log level are you using and how noisy is it ?

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        # save docs and gather metadata
+        metadata=[]
+        pdesc='Saved text to'
+        pbar=tqdm(total=self.doc_count, desc=f'{pdesc}: ...')
+        for doc_id in self.doc_ids:
+            # get saved filename and found metadata for this document
+            fn,meta = self._save_doc(doc_id)
+
+            # if we saved, update progress bar desc
+            if fn: pbar.set_description(f'{pdesc}: {fn}')
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:149]: this doesn't look like standard formatting, do you have the pre-commit hooks installed?

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        results = SolrQuerySet().facet(self.DOC_ID_FIELD, limit=self.doc_limit)
+        # store page counts and doc ids
+        self.page_counts = results.get_facets().facet_fields[self.DOC_ID_FIELD]
+        self.doc_ids = self.page_counts.keys()
+        self.doc_count = len(self.doc_ids)
+    
+    @staticmethod
+    def _get_id(doc_id:str) -> str:
+        """Method to make a file-safe version of a document ID"""
+        return doc_id.replace('/','|')
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:63]: django may have an existing escape character utility method we can use that handles more characters

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+        # limit docs queried
+        self.doc_limit = doc_limit
+
+        # subsequent paths
+        self.path_texts = os.path.join(self.path,'texts')
+        self.path_metadata = os.path.join(self.path,'metadata.json')
+        
+        # query to get initial results
+        results = SolrQuerySet().facet(self.DOC_ID_FIELD, limit=self.doc_limit)
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:54]: why use facets?

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        meta = {self.WORK_ID_FIELD:work_id, **metadata_docs[0]}
+
+        # find the pages docs
+        page_docs = [d for d in docs if d["item_type"] == "page"]
+
+        # transform into new dictionary with keys in `self.PAGE_ID_FIELD` and `self.OUTPUT_DOC_FIELDS`
+        pages = [self._transform_doc(doc,meta) for doc in page_docs]
+
+        # make sure sorted by numeric page num (i.e. "digital")
+        pages.sort(key=lambda page: page[self.PAGE_SORT_FIELD])
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:97]: Is the Solr sorting unreliable?

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            SolrQuerySet()
+            .search(**{self.DOC_ID_FIELD: doc_id})
+            .order_by(self.PAGE_ORDER_FIELD)
+        )
+
+        # populate the result cache with number of rows specified
+        docs = [
+            doc
+            for doc in result.get_results(rows=self.page_counts[doc_id])
+            if doc[self.DOC_ID_FIELD]==doc_id
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:82]: is the Solr filter unreliable?

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        ]
+
+        # find the metadata doc
+        metadata_docs = [d for d in docs if d["item_type"] == "work"]
+        assert len(metadata_docs)==1
+        meta = {self.WORK_ID_FIELD:work_id, **metadata_docs[0]}
+
+        # find the pages docs
+        page_docs = [d for d in docs if d["item_type"] == "page"]
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:91]: I'm having trouble following, why is this part needed?

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+            for key_new,key_orig in (
+                self.OUTPUT_DOC_FIELDS.items()
+            )
+        }
+
+        # return with page id
+        return {
+            self.PAGE_ID_FIELD:f'{meta[self.WORK_ID_FIELD]}_{odoc[self.PAGE_NUM_FIELD]}',
+            **odoc
+        }
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:115]: This is hard to understand. What's going on, and why is it needed? The docstring should provide more context and an example if that's needed to make this clear.

File: ppa/archive/management/commands/generate_textcorpus.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+from django.core.management.base import BaseCommand
+from parasolr.django import SolrQuerySet
+from collections import defaultdict, OrderedDict
+import logging
+from tqdm import tqdm
+from typing import Tuple
+from contextlib import contextmanager
+import logging
+
+class SolrCorpus:
```
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:18]: I vaguely remember this class from Vineet's code. It's great to keep it if you find it a useful way to encapsulate the logic. (I would probably put the methods on the manage command, personally.)

If you keep it, you should write unit tests for each method of the class.

File: ppa/archive/tests/test_generate_textcorpus.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def test_invalid_preprocess_flags(tmpdir, patched_solr_queryset):
+    # Flags that are not supported
+    with pytest.raises(CommandError):
+        call_command(
+            "generate_textcorpus", "--path", tmpdir.dirpath(), "--doc-limit","one"
+        )
+
+    with pytest.raises(CommandError):
```
Comment (rlskoeser) [inline ppa/archive/tests/test_generate_textcorpus.py:100]: fyi, `pytest.raises` has optional arguments that will match the expected error text in the exception

File: ppa/archive/management/commands/generate_textcorpus.py (hunk 4 above)
Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:54]: I think a regular search would be better than a facet. You need to add a filter on suppressed/public here. It could be helpful to use our existing ppa solr queryset object here.

File: dev-requirements.txt [hunk 11]
```diff
@@ -4,4 +4,5 @@ pytest-django>=4.5.2
 pytest-cov
 django-debug-toolbar
 sphinx
-pre-commit
\ No newline at end of file
+pre-commit
+tqdm
```
[the remark this answers is not in the document]
Comment (rlskoeser) [reply dev-requirements.txt:8]: while you're at it: could you check if we're still requiring `gensim` and comment it out? we should make it an optional install since we're not planning to run the other version of this script

**Review by rlskoeser — COMMENTED** [review]
It's really great to see a draft of this script. It looks like you were able to adapt some of Vineet's code; glad to see that. (although I haven't compared so it's hard to tell what you've changed. I do remember his version is far more complicated than we needed!)

I'm having trouble understanding some of the transformation logic - both what you're doing but more importantly why you're doing it. You should be able to get what you need out of Solr pretty directly with minimal transformation in the python code. If you need to rename the fields, you can do that when you limit which fields are returned - you can read about it in the Solr docs or look at our solr queryset object where we're doing something similar (we call it `aliases` there).

I'm also a little surprised about the number of solr queries - is it one query to get the documents and then one query per document to get the pages? Maybe this is the same as what Vineet was already doing, I don't remember, but I think we could do it more efficiently.  I'd be happy to talk through different approaches. I guess right now you're writing out one file of page content per document so maybe this is reasonable and we can just keep it in mind as a possibility for future refinement.

**Comment by rlskoeser** [comment]
@quadrismegistus have you used json lines before (JSONL), and have you considered using it for the PPA corpus?

File: ppa/archive/management/commands/generate_textcorpus.py (hunk 6 above)
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:82]: Yes, I found it was in one instance. I can try to recreate

File: ppa/archive/management/commands/generate_textcorpus.py (hunk 8 above)
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:115]: We're just returning a new output dictionary with field names changed and an additional field called page_id which is equal to {work_id}_{page_orig}.

**Comment by quadrismegistus** [comment]
> @quadrismegistus have you used json lines before (JSONL), and have you considered using it for the PPA corpus?

Yes. What benefit do you see it having in this instance? Are you imagining a .jsonl file for the entire corpus, or changing the individual work jsons to jsonls? The former seems unwieldy to me for average users and the latter seems unnecessary given that the json files are small, read in <1ms, and the jsonl file would eventually read the same file contents into memory and I'm not sure there'd by any speed improvement.

File: ppa/archive/management/commands/generate_textcorpus.py (hunk 5 above)
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:97]: I found it was in one instance, yes, I can try to recreate.

**Comment by quadrismegistus** [comment]
@rlskoeser I started a new branch/PR for a minimalist jsonl implementation of this, branching off this PR, because it kind of gets rid of a lot, but I kind of like the simplicity now: #563

I'll request review there and if you like that approach I'll merge that PR into this one.
